lightning-invoice/src/payment.rs

```diff
@@ -0,0 +1,911 @@
+//! #     fn node_id(&self) -> PublicKey { unimplemented!() }
+//! #     fn first_hops(&self) -> Vec<ChannelDetails> { unimplemented!() }
+//! #     fn send_payment(
+//! #         &self, route: &Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>
+//! #     ) -> Result<PaymentId, PaymentSendFailure> { unimplemented!() }
+//! #     fn retry_payment(
+//! #         &self, route: &Route, payment_id: PaymentId
+//! #     ) -> Result<(), PaymentSendFailure> { unimplemented!() }
+//! # }
+//! #
+//! # struct FakeRouter {};
+//! # impl Router for FakeRouter {
+//! #     fn find_route(
+//! #         &self, payer: &PublicKey, payee: &PublicKey, payee_features: Option<InvoiceFeatures>,
+//! #         first_hops: Option<&[&ChannelDetails]>, last_hops: &[&RouteHint],
+//! #         final_value_msat: u64, final_cltv: u32
+//! #     ) -> Result<Route, LightningError> { unimplemented!() }
+//! # }
+//! #
+//! # struct FakeLogger {};
+//! # impl Logger for FakeLogger {
+//! #     fn log(&self, record: &Record) { unimplemented!() }
+//! # }
+//! #
+//! # fn main() {
+//! let event_handler = |event: &Event| {
+//!     match event {
+//!         Event::PaymentPathFailed { .. } => println!("payment failed after retries"),
+//!         Event::PaymentSent { .. } => println!("payment successful"),
+//!         _ => {},
+//!     }
+//! };
+//! # let payer = FakePayer {};
+//! # let router = FakeRouter {};
+//! # let logger = FakeLogger {};
+//! let invoice_payer = InvoicePayer::new(&payer, router, &logger, event_handler, RetryAttempts(2));
+//!
+//! let invoice = "...";
+//! let invoice = invoice.parse::<Invoice>().unwrap();
+//! invoice_payer.pay_invoice(&invoice).unwrap();
+//!
+//! # let event_provider = FakeEventProvider {};
+//! loop {
+//!     event_provider.process_pending_events(&invoice_payer);
+//! }
+//! # }
+//! ```
+//!
+//! # Note
+//!
+//! The [`Route`] is computed before each payment attempt. Any updates affecting path finding such
+//! as updates to the network graph or changes to channel scores should be applied prior to
+//! retries, typically by way of composing [`EventHandler`]s accordingly.
+
+use crate::Invoice;
+
+use bitcoin_hashes::Hash;
+
+use lightning::ln::{PaymentHash, PaymentSecret};
+use lightning::ln::channelmanager::{ChannelDetails, PaymentId, PaymentSendFailure};
+use lightning::ln::features::InvoiceFeatures;
+use lightning::ln::msgs::LightningError;
+use lightning::routing::router::{Route, RouteHint};
+use lightning::util::events::{Event, EventHandler};
+use lightning::util::logger::Logger;
+
+use secp256k1::key::PublicKey;
+
+use std::collections::hash_map::{self, HashMap};
+use std::ops::Deref;
+use std::sync::Mutex;
+
+/// A utility for paying [`Invoice]`s.
+pub struct InvoicePayer<P: Deref, R, L: Deref, E>
+where
+	P::Target: Payer,
+	R: Router,
+	L::Target: Logger,
+	E: EventHandler,
+{
+	payer: P,
+	router: R,
+	logger: L,
+	event_handler: E,
+	invoice_cache: Mutex<HashMap<PaymentHash, (Invoice, usize)>>,
```

Comment:
What's the plan to adapt this to stateless-ish? We could take the lazy approach and keep this the way it is and *just* add invoice storage to the `send_payment` call, making ChannelManager/ChannelMonitor store it as an arbitrary string. Or we break it down into the list of arguments to `find_route` and `send_payment` and make that some struct in ChannelManager/Monitor. That would be pretty awkward, I think, cause then we kinda have a struct that lives in the lightning crate but then is used primarily here, but I dunno how better to do it.

Reply:
Right, that would be the second thing mentioned above - some struct that contains all the parameters to `find_route` and `send_payment` that lives in `ChannelManager` but is only serialized there (and in `ChannelMonitor`) and only *used* in lightning-invoice. I'm fine with that, just want to make sure we raise it again here now that we have more concrete code to write :).